File: data_pipeline/scale/extract-units.mjs
import crypto from "node:crypto";
import { mkdir, readFile, readdir, stat, writeFile } from "node:fs/promises";
import path from "node:path";

const args = parseArgs(process.argv.slice(2));
const checkoutRoot = path.resolve(required(args, "checkout-root"));
const repositoryManifest = path.resolve(args.repositories ?? "data_pipeline/scale/repositories.jsonl");
const output = path.resolve(args.output ?? "data_pipeline/scale/units.jsonl");
const limitPerRepo = Number(args["limit-per-repo"] ?? 36);
const targetTotal = Number(args.target ?? 500);
const repositories = parseJsonl(await readFile(repositoryManifest, "utf8")).filter((item) => item.status === "accepted");
const seenTargets = new Set();
const all = [];

for (const repository of repositories) {
  const root = path.join(checkoutRoot, repository.checkout);
  const files = await findTypeScriptFiles(root);
  const candidates = [];
  for (const filename of files) {
    const sourceText = await readFile(filename, "utf8");
    for (const extracted of extractUnits(sourceText)) {
      const relative = path.relative(root, filename).split(path.sep).join("/");
      const score = relevanceScore(`${relative} ${extracted.name} ${extracted.code}`, repository.category);
      if (score < 2) continue;
      const normalized = normalize(extracted.code);
      const digest = sha256(normalized);
      if (seenTargets.has(digest)) continue;
      const start = lineAt(sourceText, extracted.start);
      const end = lineAt(sourceText, extracted.end);
      candidates.push({ ...extracted, relative, start, end, score, digest, normalized });
    }
  }
  candidates.sort((a, b) => b.score - a.score || a.code.length - b.code.length || a.relative.localeCompare(b.relative));
  for (const unit of candidates.slice(0, limitPerRepo)) {
    seenTargets.add(unit.digest);
    all.push({
      schema_version: "1.0",
      unit_id: `scale-${slug(repository.repo)}-${String(all.length + 1).padStart(4, "0")}`,
      repository: repository.repo,
      repository_family_id: repository.family_id,
      category: repository.category,
      commit: repository.commit,
      license: repository.license.spdx_id,
      source_path: unit.relative,
      source_lines: `${unit.start}-${unit.end}`,
      symbol: unit.name,
      kind: unit.kind,
      granularity: unit.kind === "class" ? "G2" : "G1",
      signature: unit.signature,
      provided_symbols: referencedSymbols(unit.code, unit.name),
      relevance_score: unit.score,
      target_sha256: sha256(unit.code.trim()),
      target: unit.code.trim() + "\n"
    });
  }
}

const selected = all.sort((a, b) => b.relevance_score - a.relevance_score || a.unit_id.localeCompare(b.unit_id)).slice(0, targetTotal);
await mkdir(path.dirname(output), { recursive: true });
await writeFile(output, selected.map((item) => JSON.stringify(item)).join("\n") + "\n", "utf8");
const byRepo = Object.fromEntries(repositories.map((repo) => [repo.repo, selected.filter((unit) => unit.repository === repo.repo).length]));
console.log(JSON.stringify({ accepted_repositories: repositories.length, extracted: all.length, selected: selected.length, by_repo: byRepo, output }, null, 2));

function extractUnits(sourceText) {
  const result = [];
  const pattern = /(?:^|\n)([ \t]*(?:(?:export|default|declare|abstract)\s+)*(?:(?:async)\s+)?(?:function\s+([A-Za-z_$][\w$]*)\s*[^;{]*|class\s+([A-Za-z_$][\w$]*)[^;{]*)\{)/gm;
  for (const match of sourceText.matchAll(pattern)) {
    const declaration = match[1];
    const name = match[2] ?? match[3];
    const kind = match[2] ? "function" : "class";
    const declarationStart = match.index + match[0].indexOf(declaration);
    const open = declarationStart + declaration.lastIndexOf("{");
    const close = matchingBrace(sourceText, open);
    if (close < 0) continue;
    const code = sourceText.slice(declarationStart, close + 1).trim();
    const lines = code.split("\n").length;
    if (code.length < 100 || code.length > 3000 || lines < 4 || lines > 100) continue;
    if (/\b(TODO|FIXME|throw new Error\(["']not implemented)/i.test(code)) continue;
    result.push({ name, kind, signature: declaration.slice(0, declaration.lastIndexOf("{")).replace(/\s+/g, " ").trim().slice(0, 500), code, start: declarationStart, end: close });
  }
  return result;
}

function matchingBrace(text, open) {
  let depth = 0;
  let quote = null;
  let lineComment = false;
  let blockComment = false;
  for (let index = open; index < text.length; index++) {
    const char = text[index];
    const next = text[index + 1];
    if (lineComment) { if (char === "\n") lineComment = false; continue; }
    if (blockComment) { if (char === "*" && next === "/") { blockComment = false; index++; } continue; }
    if (quote) {
      if (char === "\\") { index++; continue; }
      if (char === quote) quote = null;
      continue;
    }
    if (char === "/" && next === "/") { lineComment = true; index++; continue; }
    if (char === "/" && next === "*") { blockComment = true; index++; continue; }
    if (char === "'" || char === '"' || char === "`") { quote = char; continue; }
    if (char === "{") depth++;
    if (char === "}" && --depth === 0) return index;
  }
  return -1;
}

function relevanceScore(text, category) {
  const lower = text.toLowerCase();
  const common = ["game", "player", "enemy", "turn", "round", "damage", "health", "score", "move", "position", "target", "attack", "action", "state", "random"];
  const card = ["card", "deck", "hand", "draw", "discard", "shuffle", "mana", "energy"];
  const tactics = ["grid", "tile", "path", "unit", "range", "distance", "neighbor", "battle", "combat"];
  const tower = ["tower", "wave", "projectile", "creep", "spawn", "cooldown", "upgrade", "defense"];
  let score = common.reduce((sum, word) => sum + (lower.includes(word) ? 1 : 0), 0);
  const specific = category === "card" ? card : category === "tactics" ? tactics : tower;
  score += 2 * specific.reduce((sum, word) => sum + (lower.includes(word) ? 1 : 0), 0);
  return score;
}

function referencedSymbols(code, ownName) {
  const identifiers = code.match(/[A-Za-z_$][A-Za-z0-9_$]*/g) ?? [];
  const ignored = new Set([ownName, "const", "let", "var", "function", "class", "return", "if", "else", "for", "while", "new", "this", "true", "false", "null", "undefined", "number", "string", "boolean", "void", "export", "default", "public", "private", "protected", "static", "readonly", "async", "await", "Math", "Array", "Object", "String", "Number", "Set", "Map", "Promise"]);
  return [...new Set(identifiers.filter((name) => /^[A-Z]/.test(name) && !ignored.has(name)))].slice(0, 12);
}

async function findTypeScriptFiles(root) {
  const result = [];
  async function visit(directory) {
    for (const entry of await readdir(directory, { withFileTypes: true })) {
      if (["node_modules", ".git", "dist", "build", "coverage", "vendor", "generated"].includes(entry.name)) continue;
      const item = path.join(directory, entry.name);
      if (entry.isDirectory()) await visit(item);
      else if (entry.isFile() && entry.name.endsWith(".ts") && !entry.name.endsWith(".d.ts") && !/\.(test|spec)\.ts$/.test(entry.name) && !/config/i.test(entry.name)) result.push(item);
    }
  }
  if ((await stat(root)).isDirectory()) await visit(root);
  return result;
}

function normalize(value) { return value.replace(/\/\/.*|\/\*[\s\S]*?\*\//g, "").replace(/\s+/g, " ").trim(); }
function lineAt(value, offset) { return value.slice(0, offset).split("\n").length; }
function sha256(value) { return crypto.createHash("sha256").update(value).digest("hex"); }
function slug(value) { return value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, ""); }
function parseJsonl(value) { return value.split("\n").filter(Boolean).map((line) => JSON.parse(line)); }
function required(values, name) { if (!values[name]) throw new Error(`--${name} is required`); return values[name]; }
function parseArgs(values) { const result = {}; for (let i = 0; i < values.length; i += 2) result[values[i].replace(/^--/, "")] = values[i + 1]; return result; }

File: data_pipeline/scale/extract_units.py
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path
from typing import Any, Iterable


SKIP_DIRECTORIES = {"node_modules", ".git", "dist", "build", "coverage", "vendor", "generated"}
COMMON = ("game", "player", "enemy", "turn", "round", "damage", "health", "score", "move", "position", "target", "attack", "action", "state", "random")
CATEGORY_WORDS = {
    "card": ("card", "deck", "hand", "draw", "discard", "shuffle", "mana", "energy"),
    "tactics": ("grid", "tile", "path", "unit", "range", "distance", "neighbor", "battle", "combat"),
    "tower": ("tower", "wave", "projectile", "creep", "spawn", "cooldown", "upgrade", "defense"),
}
DECLARATION = re.compile(
    r"(?:^|\n)([ \t]*(?:(?:export|default|declare|abstract)\s+)*(?:(?:async)\s+)?"
    r"(?:function\s+([A-Za-z_$][\w$]*)\s*[^;{]*|class\s+([A-Za-z_$][\w$]*)[^;{]*)\{)",
    re.MULTILINE,
)


def read_jsonl(path: Path) -> list[dict[str, Any]]:
    return [json.loads(line) for line in path.read_text(encoding="utf-8").splitlines() if line.strip()]


def write_jsonl(path: Path, rows: Iterable[dict[str, Any]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("".join(json.dumps(row, ensure_ascii=False, separators=(",", ":")) + "\n" for row in rows), encoding="utf-8")


def matching_brace(text: str, opening: int) -> int:
    depth = 0
    quote: str | None = None
    line_comment = block_comment = False
    index = opening
    while index < len(text):
        char = text[index]
        following = text[index + 1] if index + 1 < len(text) else ""
        if line_comment:
            if char == "\n":
                line_comment = False
        elif block_comment:
            if char == "*" and following == "/":
                block_comment = False
                index += 1
        elif quote:
            if char == "\\":
                index += 1
            elif char == quote:
                quote = None
        elif char == "/" and following == "/":
            line_comment = True
            index += 1
        elif char == "/" and following == "*":
            block_comment = True
            index += 1
        elif char in "'\"`":
            quote = char
        elif char == "{":
            depth += 1
        elif char == "}":
            depth -= 1
            if depth == 0:
                return index
        index += 1
    return -1


def extract_units(source: str) -> list[dict[str, Any]]:
    units: list[dict[str, Any]] = []
    for match in DECLARATION.finditer(source):
        declaration, function_name, class_name = match.group(1), match.group(2), match.group(3)
        name = function_name or class_name
        kind = "function" if function_name else "class"
        start = match.start(1)
        opening = start + declaration.rfind("{")
        closing = matching_brace(source, opening)
        if closing < 0:
            continue
        code = source[start : closing + 1].strip()
        lines = code.count("\n") + 1
        if not (100 <= len(code) <= 3000 and 4 <= lines <= 100):
            continue
        if re.search(r"\b(TODO|FIXME|throw new Error\([\"']not implemented)", code, re.IGNORECASE):
            continue
        units.append({
            "name": name,
            "kind": kind,
            "signature": re.sub(r"\s+", " ", declaration[: declaration.rfind("{")]).strip()[:500],
            "code": code,
            "start_offset": start,
            "end_offset": closing,
        })
    return units


def normalize(code: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"//.*|/\*[\s\S]*?\*/", "", code)).strip()


def sha256(value: str) -> str:
    return hashlib.sha256(value.encode()).hexdigest()


def relevance_score(text: str, category: str) -> int:
    lowered = text.lower()
    return sum(word in lowered for word in COMMON) + 2 * sum(word in lowered for word in CATEGORY_WORDS.get(category, CATEGORY_WORDS["tower"]))


def referenced_symbols(code: str, own_name: str) -> list[str]:
    ignored = {own_name, "Math", "Array", "Object", "String", "Number", "Set", "Map", "Promise"}
    return list(dict.fromkeys(value for value in re.findall(r"[A-Za-z_$][A-Za-z0-9_$]*", code) if value[:1].isupper() and value not in ignored))[:12]


def source_files(root: Path) -> list[Path]:
    return sorted(
        path for path in root.rglob("*.ts")
        if not any(part in SKIP_DIRECTORIES for part in path.parts)
        and not path.name.endswith(".d.ts")
        and not re.search(r"\.(test|spec)\.ts$", path.name)
        and "config" not in path.name.lower()
    )


def run(checkout_root: Path, repositories_path: Path, output: Path, limit_per_repo: int, target: int) -> dict[str, Any]:
    repositories = [row for row in read_jsonl(repositories_path) if row.get("status") == "accepted"]
    all_units: list[dict[str, Any]] = []
    seen: set[str] = set()
    for repository in repositories:
        root = checkout_root / repository["checkout"]
        candidates: list[dict[str, Any]] = []
        for filename in source_files(root):
            source = filename.read_text(encoding="utf-8", errors="replace")
            relative = filename.relative_to(root).as_posix()
            for unit in extract_units(source):
                score = relevance_score(f"{relative} {unit['name']} {unit['code']}", repository["category"])
                digest = sha256(normalize(unit["code"]))
                if score < 2 or digest in seen:
                    continue
                unit.update(relative=relative, score=score, digest=digest,
                            start=source.count("\n", 0, unit["start_offset"]) + 1,
                            end=source.count("\n", 0, unit["end_offset"]) + 1)
                candidates.append(unit)
        candidates.sort(key=lambda item: (-item["score"], len(item["code"]), item["relative"]))
        for unit in candidates[:limit_per_repo]:
            seen.add(unit["digest"])
            all_units.append({
                "schema_version": "1.0",
                "unit_id": f"scale-{re.sub(r'(^-|-$)', '', re.sub(r'[^a-z0-9]+', '-', repository['repo'].lower()))}-{len(all_units)+1:04d}",
                "repository": repository["repo"], "repository_family_id": repository["family_id"],
                "category": repository["category"], "commit": repository["commit"],
                "license": repository["license"]["spdx_id"], "source_path": unit["relative"],
                "source_lines": f"{unit['start']}-{unit['end']}", "symbol": unit["name"], "kind": unit["kind"],
                "granularity": "G2" if unit["kind"] == "class" else "G1", "signature": unit["signature"],
                "provided_symbols": referenced_symbols(unit["code"], unit["name"]), "relevance_score": unit["score"],
                "target_sha256": sha256(unit["code"].strip()), "target": unit["code"].strip() + "\n",
            })
    selected = sorted(all_units, key=lambda item: (-item["relevance_score"], item["unit_id"]))[:target]
    write_jsonl(output, selected)
    return {"accepted_repositories": len(repositories), "extracted": len(all_units), "selected": len(selected),
            "by_repo": {repo["repo"]: sum(unit["repository"] == repo["repo"] for unit in selected) for repo in repositories},
            "output": str(output)}


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--checkout-root", type=Path, required=True)
    parser.add_argument("--repositories", type=Path, default=Path("data_pipeline/scale/repositories.jsonl"))
    parser.add_argument("--output", type=Path, default=Path("data_pipeline/scale/units.jsonl"))
    parser.add_argument("--limit-per-repo", type=int, default=36)
    parser.add_argument("--target", type=int, default=500)
    args = parser.parse_args()
    print(json.dumps(run(args.checkout_root.resolve(), args.repositories.resolve(), args.output.resolve(), args.limit_per_repo, args.target), indent=2))


if __name__ == "__main__":
    main()
